docker.container regression

```shell
$ ./converge apply --local samples/dockerContainer.hcl
2016-10-03T11:17:27-04:00 |INFO| serving        addr=:47740 component=rpc
2016-10-03T11:17:27-04:00 |WARN| skipping module verification   component=client
2016-10-03T11:17:27-04:00 |INFO| got status     component=client file=samples/dockerContainer.hcl id=root/docker.container.nginx run=STARTED stage=APPLY
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x233635]

goroutine 78 [running]:
panic(0x608380, 0xc4200160c0)
        /usr/local/go/src/runtime/panic.go:500 +0x1a1
github.com/asteris-llc/converge/apply.(*pipelineGen).applyNode(0xc4202b0140, 0x61d0c0, 0xc4202a75f0, 0x61d0c0, 0xc4202a75f0, 0x0, 0x0)
        /Users/ryan/Projects/golang/src/github.com/asteris-llc/converge/apply/pipeline.go:135 +0x145
github.com/asteris-llc/converge/apply.(*pipelineGen).(github.com/asteris-llc/converge/apply.applyNode)-fm(0x61d0c0, 0xc4202a75f0, 0x61d0c0, 0xc4202a75f0, 0x0, 0x0)
        /Users/ryan/Projects/golang/src/github.com/asteris-llc/converge/apply/pipeline.go:45 +0x3e
github.com/asteris-llc/converge/executor.Pipeline.Exec(0xc4202a2d40, 0x8, 0x8, 0x617e40, 0xc42029cbe0, 0x8, 0x63b260, 0xc420291f80, 0x0)
        /Users/ryan/Projects/golang/src/github.com/asteris-llc/converge/executor/pipeline.go:51 +0x5a
github.com/asteris-llc/converge/apply.execPipeline.func1(0xc4201453e0, 0x1b, 0xc4202a6e70, 0x0, 0x1b)
        /Users/ryan/Projects/golang/src/github.com/asteris-llc/converge/apply/apply.go:71 +0x10a
github.com/asteris-llc/converge/graph.NotifyPre.func1(0xc4201453e0, 0x1b, 0xc4202a6e70, 0xc4202919c8, 0xc420299cb0)
        /Users/ryan/Projects/golang/src/github.com/asteris-llc/converge/graph/notifier.go:30 +0x7f
github.com/asteris-llc/converge/graph.NotifyPost.func1(0xc4201453e0, 0x1b, 0xc4202a6e70, 0xc42029cbe0, 0xc420299e01)
        /Users/ryan/Projects/golang/src/github.com/asteris-llc/converge/graph/notifier.go:37 +0x54
github.com/asteris-llc/converge/graph.transform.func1(0xc4201453e0, 0x1b, 0x617e40, 0xc42029cbe0, 0xc42029cbe0, 0xc4202a2940)
        /Users/ryan/Projects/golang/src/github.com/asteris-llc/converge/graph/graph.go:554 +0x45
github.com/asteris-llc/converge/graph.dependencyWalk.func5(0xc4201453e0, 0x1b)
        /Users/ryan/Projects/golang/src/github.com/asteris-llc/converge/graph/graph.go:349 +0xb2c
created by github.com/asteris-llc/converge/graph.dependencyWalk.func3
        /Users/ryan/Projects/golang/src/github.com/asteris-llc/converge/graph/graph.go:264 +0x4f6
```

--- a/apply/pipeline.go
+++ b/apply/pipeline.go
@@ -15,7 +15,6 @@
 package apply
 
 import (
-	"errors"
 	"fmt"
 
 	"github.com/asteris-llc/converge/executor"
@@ -23,6 +22,7 @@ import (
 	"github.com/asteris-llc/converge/plan"
 	"github.com/asteris-llc/converge/render"
 	"github.com/asteris-llc/converge/resource"
+	"github.com/pkg/errors"
 )
 
 type pipelineGen struct {
@@ -162,7 +162,11 @@ func (g *pipelineGen) maybeRunFinalCheck(resultI interface{}) (interface{}, erro
 	}
 	result.PostCheck = planned.Status
 	if planned.HasChanges() {
-		result.Err = fmt.Errorf("%s still has changes after apply", g.ID)
+		if result.Err != nil {
+			result.Err = errors.Wrap(result.Err, fmt.Sprintf("%s still has changes after apply", g.ID))
+		} else {
+			result.Err = fmt.Errorf("%s still has changes after apply", g.ID)
+		}
 	}
 	return result, nil
 }

--- a/resource/docker/container/container.go
+++ b/resource/docker/container/container.go
@@ -60,9 +60,7 @@ type Container struct {
 
 // Check that a docker container with the specified configuration exists
 func (c *Container) Check(resource.Renderer) (resource.TaskStatus, error) {
-	if c.Status == nil {
-		c.Status = resource.NewStatus()
-	}
+	c.Status = resource.NewStatus()
 	container, err := c.client.FindContainer(c.Name)
 	if err != nil {
 		c.Status.Level = resource.StatusFatal
@@ -87,9 +85,7 @@ func (c *Container) Check(resource.Renderer) (resource.TaskStatus, error) {
 
 // Apply starts a docker container with the specified configuration
 func (c *Container) Apply() (resource.TaskStatus, error) {
-	if c.Status == nil {
-		c.Status = resource.NewStatus()
-	}
+	c.Status = resource.NewStatus()
 	volumes, binds := volumeConfigs(c.Volumes)
 	config := &dc.Config{
 		Image:        c.Image,

--- a/resource/docker/container/container.go
+++ b/resource/docker/container/container.go
@@ -60,7 +60,9 @@ type Container struct {
 
 // Check that a docker container with the specified configuration exists
 func (c *Container) Check(resource.Renderer) (resource.TaskStatus, error) {
-	c.Status = resource.NewStatus()
+	if c.Status == nil {
+		c.Status = resource.NewStatus()
+	}
 	container, err := c.client.FindContainer(c.Name)
 	if err != nil {
 		c.Status.Level = resource.StatusFatal
@@ -85,6 +87,9 @@ func (c *Container) Check(resource.Renderer) (resource.TaskStatus, error) {
 
 // Apply starts a docker container with the specified configuration
 func (c *Container) Apply() (resource.TaskStatus, error) {
+	if c.Status == nil {
+		c.Status = resource.NewStatus()
+	}
 	volumes, binds := volumeConfigs(c.Volumes)
 	config := &dc.Config{
 		Image:        c.Image,
@@ -113,13 +118,13 @@ func (c *Container) Apply() (resource.TaskStatus, error) {
 
 	container, err := c.client.CreateContainer(opts)
 	if err != nil {
-		return nil, errors.Wrapf(err, "failed to run container %s", c.Name)
+		return c, err
 	}
 
 	if c.CStatus == "" || c.CStatus == containerStatusRunning {
 		err = c.client.StartContainer(c.Name, container.ID)
 		if err != nil {
-			return nil, errors.Wrapf(err, "failed to start container %s", c.Name)
+			return c, err
 		}
 	}
 

--- a/resource/docker/container/preparer.go
+++ b/resource/docker/container/preparer.go
@@ -120,6 +120,12 @@ func (p *Preparer) Prepare(render resource.Renderer) (resource.Task, error) {
 }
 
 func validateContainer(container *Container) error {
+	if container.Name == "" {
+		return errors.New("name must be provided")
+	}
+	if container.Image == "" {
+		return errors.New("image must be provided")
+	}
 	if container.CStatus != "" {
 		if !strings.EqualFold(container.CStatus, containerStatusRunning) &&
 			!strings.EqualFold(container.CStatus, containerStatusCreated) {

--- a/resource/docker/container/preparer_test.go
+++ b/resource/docker/container/preparer_test.go
@@ -23,15 +23,36 @@ import (
 	"github.com/stretchr/testify/assert"
 )
 
+// TestPreparerInterface tests that the Preparer interface is properly
+// implemented
 func TestPreparerInterface(t *testing.T) {
 	t.Parallel()
 	assert.Implements(t, (*resource.Resource)(nil), new(container.Preparer))
 }
 
+// TestPreparerInvalidStatus tests preparer validation
 func TestPreparerInvalidStatus(t *testing.T) {
-	p := &container.Preparer{Name: "test", Image: "nginx", Status: "exited"}
-	_, err := p.Prepare(fakerenderer.New())
-	if assert.Error(t, err) {
-		assert.EqualError(t, err, "status must be 'running' or 'created'")
-	}
+	t.Run("status is invalid", func(t *testing.T) {
+		p := &container.Preparer{Name: "test", Image: "nginx", Status: "exited"}
+		_, err := p.Prepare(fakerenderer.New())
+		if assert.Error(t, err) {
+			assert.EqualError(t, err, "status must be 'running' or 'created'")
+		}
+	})
+
+	t.Run("name is invalid", func(t *testing.T) {
+		p := &container.Preparer{Name: "", Image: "nginx"}
+		_, err := p.Prepare(fakerenderer.New())
+		if assert.Error(t, err) {
+			assert.EqualError(t, err, "name must be provided")
+		}
+	})
+
+	t.Run("image is invalid", func(t *testing.T) {
+		p := &container.Preparer{Name: "nginx", Image: ""}
+		_, err := p.Prepare(fakerenderer.New())
+		if assert.Error(t, err) {
+			assert.EqualError(t, err, "image must be provided")
+		}
+	})
 }

